docs/rfc/rfc-001-end-to-end-testing.md

```diff
@@ -0,0 +1,137 @@
+
+All of the settings should be specified in a testnet configuration (or alternatively the seed that generated it) such that it can be retrieved from CI and debugged locally.
+
+### Test Stages
+
+Given a test configuration, the test runner has the following stages:
+
+- Setup: configures the Docker containers and networks, but does not start them.
+
+- Initialization: starts the Docker containers, performs fast sync/state sync, adds/removes validators, restarts nodes, perturbs networking, etc. Between each operation, check basic liveness and readiness.
+
+- Test: runs RPC tests independently against all network nodes, making sure data matches expectations and invariants hold.
+
+### Tests
+
+The test suite will do black-box testing against the Tendermint RPC service of a single network node, executed against all nodes separately. We will be testing the behavior of the network as a whole, e.g. that a fast synced node correctly catches up to the chain head and serves basic block data via RPC. Comprehensive component testing (e.g. each and every RPC method parameter) should be done via unit/integration tests. Thus the tests will not send e.g. P2P messages or examine the node database, as these are considered internal implementation details - if the network behaves correctly, presumably the internal components function correctly.
+
+The general approach will be to put the network through a sequence of operations (see initialization stage above), check basic liveness and readiness after each operation, and then once the network stabilizes run the test suite against each node in the network.
+
+The tests must take into account the node configuration (e.g. some nodes may be pruned, others may not be validators), and should somehow be provided access to expected data (i.e. complete block headers for the entire chain).
+
+The test suite should use the Tendermint RPC client and the Tendermint light client, to exercise the client code as well.
+
+### Implementation Considerations
+
+The testnets should run in Docker Compose, both locally and in CI. This makes it easier to reproduce test failures locally. Supporting multiple test-runners (e.g. on VMs or Kubernetes) is out of scope. The same image should be used for all tests, with configuration passed via a mounted volume.
+
+There does not appear to be any off-the-shelf solutions that would do this for us, so we will have to roll our own on top of Docker Compose. This gives us more flexibility, but is estimated to be a few weeks of work.
+
+Testnets should be configured via a YAML file. These are used as inputs for the test runner, which e.g. generates Docker Compose configurations from them. An additional layer on top should generate these testnet configurations from a YAML file that specifies all the option combinations to test.
```

Comment:
Why not supply config.toml and genesis.json files directly? Because they are too big and you want to write some kind of generator (from yaml to config.toml/genesis.json)?

Reply:
The config.toml file must be different for different nodes - both because we want to vary setting per node, and also because e.g. p2p settings are different for every node. Providing them all is too unwieldy, so they must be generated.